Fix tfuser and tffarmer with new seed option

maxux opened the pull request:
- Add tffarmer custom seed field
- Fix tfuser to create new seed without specifying a seed existing
- Fix seed needed twice for provision #646
- Improve how seed is required or not on tfuser
- Support seed 1.1.0 which bundle seed mnemonic and threebot id
- Update `tffarmer` to support new seed and use threebot id from it
- Update `tfuser` to use new seed and use threebot id from seed
- Move `tfuser id` to `tfuser id create` command
- Add `tfuser id show` and `tfuser id convert` commands

Seed argument is not mandatory but can be specified globally. If the seed is mandatory for a subcommand, it will be rejected if not available.

maxux committed:
Try to fix tfuser id

maxux committed:
Allow custom seed file for tffarmer

maxux committed:
Make tfuser seed argument required for specific actions

You can now pass --seed for any subcommands of tfuser, but only
subcommands which really need it will make it mandatory.

More subcommands maybe will need them, this could be fix later. This fix
issue #660 and fix bugs on tfuser (eg: could not create new seed)

maxux committed:
Update seeds version

maxux committed:
Update go modules

maxux committed:
Add support for seed 1.1.0 on identity package

maxux committed:
Add seedconverter tool

maxux committed:
Update how seedconv pk are compared

maxux committed:
Hide mnemonic when saving it

maxux committed:
Fix lint issues

maxux committed:
Update tfuser to create new identity seed

maxux committed:
Split identity seed loader function

maxux committed:
Support tfuser id generated from existing mnemonic

maxux committed:
Update tffarmer to use new seed features

maxux committed:
Merge branch 'master' into tools-seed

maxux committed:
Fix lint

maxux committed:
Stick old seed loader with version 1.0.0

zaibon commented on pkg/identity/userid.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+// UserData contains all information needed to identify someone
+type UserData struct {
```
This struct could be avoided I think. You already have all the information required in UserIdentity. Instead of having another struct just to have access to the KeyPair, you can just add a KeyPair method on the UserIdentity struct.

maxux replied:
Ah, yes. Wasn't sure we could do that in Go, it's much better, for sure. I'll change that.

zaibon commented on pkg/identity/userid.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+// LoadUserIdentityObject converts a UserIdentity object to UserData
+func LoadUserIdentityObject(ui *UserIdentity) (UserData, error) {
```
This can also be removed if you add the KeyPair method on UserIdentity struct

maxux replied:
Yep, cool

zaibon commented on pkg/identity/userid.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+// SaveUserData dumps UserData into a file
+func SaveUserData(u *UserData, path string) error {
```
Same here.

maxux replied:
This can be set as a method, right ?

zaibon replied:
Well the UserData struct won't exist anymore so this method has no reason o exists

zaibon commented on tools/seedconv/main.go:
```diff
@@ -0,0 +1,52 @@
+package main
```
Instead of add a new CLI move this into a sub command of `tfuser id`

maxux replied:
True, was an external tool to make things easier to test, I can now move it.

zaibon commented on tools/tfuser/cmds_identity.go:
```diff
@@ -5,46 +5,78 @@ import (
-		fmt.Printf("a seed already exists at %s\n", output)
-		fmt.Printf("identity: %s\n", k.Identity())
+		fmt.Printf("A seed already exists at: %s\n", output)
+		fmt.Printf("Identity: %s\n", eud.Key.Identity())
```
while we are at it, let's also print the threebot ID here.

zaibon commented on tools/tfuser/main.go:
```diff
@@ -6,17 +6,27 @@ import (
 )
 
-func main() {
+func requireSeed(c *cli.Context) error {
```
This is not needed, you can set the `Required` property of the `cli.StringFlag` to true on line 50 and have the same result.

maxux replied:
No, this was the initial issue. It you set it as required, you NEED to specify it for ALL subcommands, and you can't specify it for `id` subcommand. This is the exact problem of the main issue :p

zaibon replied:
Right 👌

maxux committed:
Remove UserData and use UserIdentity for everything

Remove identity UserData struct and use UserIdentity struct for all
operations with methods on it.

maxux committed:
Update syntax to load UserIdentify from existing mnemonic

maxux committed:
Notify old seed file detected

maxux committed:
Embed 'seedconv' to tfuser and improve tfuser with new seed

Update tfuser id commands:
  tfuser id show: dump user.seed information
  tfuser id create: create a new user.seed
  tfuser id convert: convert and old user.seed to new format

maxux committed:
Deprecate identity.LoadKeyPair function

maxux committed:
Improve LoadKeyPair deprecation

maxux committed:
Load keypair with legacy way when converting id

maxux committed:
Update tfuser to use threebot id from user seed

maxux committed:
Save old keypair using old seed version

maxux committed:
Update deprecated function comment

zaibon commented on pkg/identity/userid.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+// Key returns the internal KeyPair
+func (u *UserIdentity) Key() KeyPair {
```
Instead of having a method to change the key, I would have a constructor instead. So you still have a way to build a `UserIdentity` from a specific key pair. But once the object is created you cannot mess it up by changing the key.

It seems this function is only used here https://github.com/threefoldtech/zos/pull/673/files#diff-4b63591c005a6e0aa77c5d468ba7b616R99

zaibon commented on tools/tfuser/cmds_identity.go:
```diff
@@ -1,50 +1,141 @@
+	// Create new object
+	ui := &identity.UserIdentity{}
+	ui.SetKey(kp)
+	ui.ThreebotID = uint64(tid)
```
```suggestion
	ui := identity.NewUserIdentity(pk, tid)
```

maxux committed:
Disable UserIdentity SetKey and add new constructor

maxux committed:
Merge branch 'master' into tools-seed

maxux committed:
Fix UserIdentity constructor
